### PR TITLE
Add initial support for backward tracing

### PR DESCRIPTION
**Pull request by apaszke**
**This works only for single backwards and breaks in-place.**

I removed the fine-grained hooks and simplified the code. I think we'll need to take a completely different approach if we want to solve this properly...

I don't know why but `EliminateDeadCode` segfaults on my mac when I'm dereferencing the iterator. Trace passes lint, same code work with gcc, I can't find any UB in how I handle the iterators.

### CODE REVIEW

**Comment by ezyang** [comment]
If I had to give a wild guess about the iterator, it's because reverse iterators are special.

**Comment by ezyang** [comment]
Look, here is a minimal repro which segfaults on reverse iterators, but not on regular ones:

```
#include <list>
struct Node {
    std::list<Node*>::iterator iter_;
};
int main() {
    Node n1;
    Node n2;
    Node n3;
    std::list<Node*> xs;
    n1.iter_ = xs.insert(xs.end(), &n1);
    n2.iter_ = xs.insert(xs.end(), &n2);
    n3.iter_ = xs.insert(xs.end(), &n3);
    for (auto it = xs.rbegin(); it != xs.rend();) {
        Node* x = *it++;
        xs.erase(x->iter_);
        x->iter_ = xs.end();
    }
    return 0;
}
```

**Comment by ezyang** [comment]
And here's the reason: recall that reverse iterators physically point to +1 of the element they logically point to. So even though you `it++` before erasing to avoid invalidating the iterator, in fact, `it + 1` is referencing the element you are about to erase, invalidating the reverse iterator. This SO question (https://stackoverflow.com/questions/11339516/stdlist-reverse-iterating-erasing-causes-crash) suggests you should convert a reverse iterator into a forward one, delete it, then convert it back, but this seems dumb to me.

It seems to me that @zdevito was right and we shouldn't have used `std::list`, and we should bring back the manually linked version, or write our own version of `std::list` which  has a valid before-beginning iterator so that we can make the reverse iterator not insane.

File: test/test_jit.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def test_lstm(self):
         torch._C._jit_pass_lint(trace)
         self.assertExpected(str(trace))
 
+    @unittest.skip("in-place is not supported")
```
Comment (ezyang) [inline test/test_jit.py:42]: Might be helpful to just delete the `inplace=True` bits, then we still have a valid test.

File: torch/csrc/autograd/function.cpp [hunk 2]
```diff
@@ -36,27 +37,49 @@ auto Function::name() -> std::string {
 // This function is analogous to make_trace which operates on PythonOp, but this
 // function instead works for C++ implemented autograd Functions, which don't
 // actually have any backing Python class. We still need to trace them!
-void Function::createTrace(const variable_list& inputs, const variable_list& outputs) {
+variable_list Function::tracedApply(variable_list inputs) {
   using namespace torch::jit;
+  bool is_traceable = static_cast<bool>(dynamic_cast<Identity*>(this));
```
Comment (ezyang) [inline torch/csrc/autograd/function.cpp:42]: This particular test is hardcoded atm and will be generalized, IIUC.

File: torch/csrc/autograd/function.cpp [hunk 3]
```diff
@@ -36,27 +37,49 @@ auto Function::name() -> std::string {
...
+  // Traceable Functions are completely transparent to the JIT.
+  if (is_traceable) {
+    return apply(inputs);
+  }
   auto state = tracer::getTracingState(inputs);
+
+  // Register eval hooks if backward of this function is not traceable.
+  // This has to be done early, because it modifies inputs.
+  bool is_backward_traceable = false;
```
Comment (ezyang) [inline torch/csrc/autograd/function.cpp:51]: Ditto.

File: test/test_jit.py [hunk 4]
```diff
@@ -136,18 +138,38 @@ def forward(self, x):
...
+
+        x = a
+        y = a * b
+
+        trace, (x, y) = torch._C._tracer_enter((x, y))
+        z = y * 2 * x
+        z, = torch._C._tracer_exit((z,))
+        torch._C._jit_pass_lint(trace)
+
+        grad, = torch.autograd.grad(z, x, Variable(torch.ones(2, 2), requires_grad=True), create_graph=True)
```
Comment (ezyang) [inline test/test_jit.py:168]: It would be good to show that `backward()` also works

File: torch/csrc/jit/ir.h [hunk 5]
```diff
@@ -258,7 +259,7 @@ struct Node {
     return type()->kind() == TypeKind::MultiType;
   }
   void setType(const Type* type) {
-    type_ = std::move(type->clone());
+    type_ = type->clone();
```
Comment (ezyang) [inline torch/csrc/jit/ir.h:262]: Thanks :)

File: torch/csrc/autograd/function.cpp [hunk 6]
```diff
@@ -36,27 +37,49 @@ auto Function::name() -> std::string {
...
-  std::vector<Node*> value_traces;
-  value_traces.reserve(inputs.size());
+  auto* this_node = graph->create<CppOp>(getSharedPtr());
   for (auto& input: inputs) {
-    value_traces.emplace_back(tracer::getValueTrace(state, input));
-  }
-  auto* this_node = graph->appendNewNode<CppOp>(getSharedPtr());
-  for (auto value: value_traces) {
-    this_node->addInput(value);
+      this_node->addInput(tracer::getValueTrace(state, input));
```
Comment (ezyang) [inline torch/csrc/autograd/function.cpp:61]: Botched merge? Won't this run afoul of "See Note [getValueTrace can allocate nodes]"?

File: torch/csrc/autograd/function.cpp [hunk 7]
```diff
@@ -36,27 +37,49 @@ auto Function::name() -> std::string {
...
+      sel->inferTypeFrom(output->data);
+      tracer::setValueTrace(state, output, sel);
+  }
+
+  // Register the point where Eval region starts in backward.
+  // NOTE: this modifies outputs.
+  if (!is_backward_traceable) {
+    tracer::EvalEnterHook::registerHook(state, outputs, eval_state);
   }
+  return outputs;
```
Comment (ezyang) [inline torch/csrc/autograd/function.cpp:82]: It is too bad there is no way to easily deduplicate this logic with the Python verison...

File: torch/csrc/autograd/python_function.cpp [hunk 8]
```diff
@@ -652,27 +651,68 @@ PyObject* process_outputs(THPFunction* grad_fn, const UnpackedInput& unpacked, T
...
+  info.tracing_state = tracer::getTracingState(unpacked.input_vars);
+
+
+  // TODO: actually trace backward of some ops (e.g. Add)
+  info.is_backward_traceable = false;
+  if (!info.is_backward_traceable) {
+    // NOTE: this modifies unpacked.input_vars
+    info.eval_state = tracer::EvalExitHook::registerHook(info.tracing_state, unpacked.input_vars);
+
+    // Make a copy of raw_inputs with new Variables
```
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp:677]: Irritating, but I see why this is needed...

**Review by ezyang — APPROVED** [review]
Main components of this patch:

* New hook infrastructure for replacing non-traceable but possibly autograd-using segments of code with Eval nodes. Lives in tracer.cpp

* Code to hook this up with CppOp (autograd/function.cpp)

* Code to hook this up with PythonOp (autograd/python_function.cpp), with the added complication that we have to schlep everything back to PyObject

* A new DCE pass

The way I read the patch was I first skimmed the new hook additions, then read the CppOp code, compared it against the PythonOp code, and then re-read the hook additions again.

**Comment by zdevito** [comment]
merging so that we can continue with end-to-end example.

**Pull request merged**